gtsam/hybrid/HybridFactorGraph.cpp

```diff
@@ -0,0 +1,174 @@
+  for (auto &k : frontalKeys) {
+    if (discreteCardinalities.find(k) != discreteCardinalities.end()) {
+      discreteFrontals.insert(discreteCardinalities.at(k));
+    } else {
+      continuousFrontals.insert(k);
+    }
+  }
+
+  // Fill in discrete frontals and continuous frontals for the end result
+  for (auto &k : separatorKeys) {
+    if (discreteCardinalities.find(k) != discreteCardinalities.end()) {
+      discreteSeparator.insert(discreteCardinalities.at(k));
+    } else {
+      continuousSeparator.insert(k);
+    }
+  }
+
+  std::cout << RED_BOLD << "Keys: " << RESET;
+  for (auto &f : frontalKeys) {
+    if (discreteCardinalities.find(f) != discreteCardinalities.end()) {
+      auto &key = discreteCardinalities.at(f);
+      std::cout << boost::format(" (%1%,%2%),") % DefaultKeyFormatter(key.first) % key.second;
+    } else {
+      std::cout << " " << DefaultKeyFormatter(f) << ",";
+    }
+  }
+
+  if (separatorKeys.size() > 0) {
+    std::cout << " | ";
+  }
+
+  for (auto &f : separatorKeys) {
+    if (discreteCardinalities.find(f) != discreteCardinalities.end()) {
+      auto &key = discreteCardinalities.at(f);
+      std::cout << boost::format(" (%1%,%2%),") % DefaultKeyFormatter(key.first) % key.second;
+    } else {
+      std::cout << DefaultKeyFormatter(f) << ",";
+    }
+  }
+  std::cout << "\n" << RESET;
+  // PRODUCT: multiply all factors
+  gttic(product);
+
+  HybridConditional sum(KeyVector(continuousSeparator.begin(), continuousSeparator.end()),
+                        DiscreteKeys(discreteSeparator.begin(), discreteSeparator.end()),
+                        separatorKeys.size());
+
+  //  HybridDiscreteFactor product(DiscreteConditional());
+  //  for (auto&& factor : factors) product = (*factor) * product;
+  gttoc(product);
+
+  // sum out frontals, this is the factor on the separator
+  gttic(sum);
+  //  HybridFactor::shared_ptr sum = product.sum(frontalKeys);
+  gttoc(sum);
+
+  // Ordering keys for the conditional so that frontalKeys are really in front
+  Ordering orderedKeys;
+  orderedKeys.insert(orderedKeys.end(), frontalKeys.begin(),
+                     frontalKeys.end());
+  orderedKeys.insert(orderedKeys.end(), sum.keys().begin(),
+                     sum.keys().end());
+
+  // now divide product/sum to get conditional
+  gttic(divide);
+  //  auto conditional =
+  //      boost::make_shared<HybridConditional>(product, *sum, orderedKeys);
+  gttoc(divide);
+
+  auto conditional = boost::make_shared<HybridConditional>(
+      CollectKeys({continuousFrontals.begin(), continuousFrontals.end()},
+                  {continuousSeparator.begin(), continuousSeparator.end()}),
+      CollectDiscreteKeys({discreteFrontals.begin(), discreteFrontals.end()},
+                          {discreteSeparator.begin(), discreteSeparator.end()}),
+      continuousFrontals.size() + discreteFrontals.size());
+  std::cout << GREEN_BOLD << "[Conditional]\n" << RESET;
```

Comment:
printing should be removed in PR. Add back in when debugging

Reply:
I think now it's better to have them there so you can see what is going on just by running the branch locally.

Reply:
They are now guarded by `if(DEBUG)`